src/hmem_xpmem.c

```diff
@@ -0,0 +1,237 @@
+			"Whether to use XPMEM over CMA when possible "
+			"(default: no)");
+	fi_param_define(&core_prov, "xpmem_memcpy_chunksize", FI_PARAM_SIZE_T,
+			"Maximum size used for a single memcpy call"
+			"(default: %d)", XPMEM_DEFAULT_MEMCPY_CHUNK_SIZE);
+
+	xpmem = calloc(sizeof(*xpmem), 1);
+	if (!xpmem)
+		return -FI_ENOMEM;
+
+	fh = fopen("/proc/self/maps", "r");
+	if (NULL == fh) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+				"could not open /proc/self/maps for reading");
+		ret = -FI_EIO;
+		goto fail;
+	}
+
+	while (fgets(buffer, sizeof(buffer), fh)) {
+		uintptr_t low, high;
+		char *tmp;
+		/* each line of /proc/self/maps starts with low-high in
+		 * hexidecimal (without a 0x) */
+		low = strtoul(buffer, &tmp, 16);
+		high = strtoul(tmp + 1, NULL, 16);
+		if (address_max < high) {
+			address_max = high;
+		}
+	}
+	fclose(fh);
+
+	if (address_max == 0) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"could not determine the address max");
+		ret = -FI_EIO;
+		goto fail;
+	}
+
+	/* save the calcuated maximum */
+	xpmem->pinfo.address_max = address_max - 1;
+
+	/* export the process virtual address space for use with xpmem */
+	xpmem->pinfo.seg_id = xpmem_make(0, XPMEM_MAXADDR_SIZE, XPMEM_PERMIT_MODE,
+					 (void *) 0666);
+	if (xpmem->pinfo.seg_id == -1) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Failed to export process virtual address space for use with xpmem\n");
+		ret = -FI_ENODATA;
+		goto fail;
+	}
+
+	ret = fi_param_get_size_t(&core_prov, "xpmem_memcpy_chunksize",
+				  &xpmem->memcpy_chunk_size);
+	if (ret)
+		xpmem->memcpy_chunk_size = XPMEM_DEFAULT_MEMCPY_CHUNK_SIZE;
+
+	return 0;
+
+fail:
+	free(xpmem);
+	xpmem = NULL;
+	return ret;
+}
+
+int xpmem_cleanup(void)
+{
+	int ret = 0;
+
+	if (xpmem)
+		ret = xpmem_remove(xpmem->pinfo.seg_id);
+
+	return (ret) ? -FI_EINVAL : ret;
+}
+
+int xpmem_get_handle_size(size_t *size)
+{
+	*size = sizeof(struct xpmem_addr);
+	return FI_SUCCESS;
+}
+
+int xpmem_open_handle(void **handle, size_t len, uint64_t device, void **ipc_ptr)
+{
+	struct xpmem_addr *xpmem_addr = (struct xpmem_addr *) handle;
+
+	*ipc_ptr = xpmem_attach(*xpmem_addr, len, NULL);
+	if (*ipc_ptr == (void *) -1)
+		return -FI_EIO;
+
+	return 0;
+}
+
+int xpmem_close_handle(void *ipc_ptr)
+{
+	return xpmem_detach(ipc_ptr);
+}
+
+static inline void xpmem_memcpy(void *dst, void *src, size_t size)
+{
+	while (size > 0) {
+		size_t copy_size = MIN(size, xpmem->memcpy_chunk_size);
+		memcpy(dst, src, copy_size);
+		dst = (void *) ((uintptr_t) dst + copy_size);
+		src = (void *) ((uintptr_t) src + copy_size);
+		size -= copy_size;
+	}
+}
+
+int xpmem_copy_from(uint64_t device, void *dst, const void *src,
+		       size_t size)
+{
+	xpmem_memcpy(dst, (void *)src, size);
+	return 0;
+}
+
+int xpmem_copy_to(uint64_t device, void *dst, const void *src,
+		     size_t size)
+{
+	xpmem_memcpy(dst, (void *)src, size);
+	return 0;
+}
+
+bool xpmem_is_addr_valid(const void *addr, uint64_t *device, uint64_t *flags)
+{
+	return false;
+}
```

Comment:
Always returning false seems odd.  How would the xpmem path ever be taken?

Reply:
To follow up on the Nov 29, 2022 meeting discussion:

I would argue that XPMEM is another variation of the HMEM_SYSTEM. 
In fact we could have something like HMEM_SYSTEM_XPMEM, HMEM_SYSTEM_CMA, etc
The HMEM_SYSTEM option behaves in the same way as XPMEM with regards to the is_addr_valid() API. In fact it doesn't define it at all, because it assumes it'll never be called. It's explicitly checked in ofi_get_hmem_iface(). It also sets a bunch of the other callbacks to the ofi_no_...() version.

It seems that the entire infrastructure was created with the assumption that not all interfaces will support all the different callbacks. So it's normal to have an interface not support some of the callbacks.

If we look at SHM usage, when it selects the iface with smr_get_mr_hmem_iface() it can return FI_HMEM_SYSTEM.
This could result in both peers using the FI_HMEM_SYSTEM for H2H runs. In that case it'll go through the HMEM API. Thinking about it some more here, it's possible to return XPMEM if XPMEM is enabled and the user indicated they want to use it.

Given the way it's currently used, it makes the argument that XPMEM option fits within this infrastructure, with the limitation of having it not exposed to the external API.

With regards to the is_addr_valid() call, I do think it's a bug to just return false. I think there is a way to verify if the pointer is in the area mapped by XPMEM (have to investigate).

The point I'm trying to make here is that the reason I'm not convinced of separating XPMEM from HMEM, is not only because of code duplication, but because I believe it's a fit within the existing infrastructure for the reasons I mentioned above

Reply:
@shefty, @aingerson 
Just adding tags to make sure notifications are generated for my above comment.